### ISSUE TITLE
Random Forest Quantile Regression

### ISSUE THREAD

**Issue by mhoward2718** [microsoft/LightGBM#2607]
## Summary

LightGBM could do quantile regression in a different way through random forest mode. 

## Motivation

With the random forest method, a single model can give estimates for any quantile, versus the existing approach which requires specification of a target quantile at training time.

## Description

I'm unfamiliar with the internals of LightGBM, but this might be a pretty low effort feature. As I see it, the two required pieces are:

- Storage of weights/values in leaves as described below
- Specification of quantile (alpha) in predict method

Relevant description copy pasted from the reference below: 

To estimate F(Y=y|x)=q each target value in y_train is given a weight. Formally, the weight given to y_train[j] while estimating the quantile is 1T∑Tt=1𝟙(yj∈L(x))∑Ni=1𝟙(yi∈L(x)) where L(x) denotes the leaf that x falls into.

Informally, what it means that for a new unknown sample, we first find the leaf that it falls into at each tree. Then for each (X, y) in the training data, a weight is given to y at each tree in the following manner.

If it is in the same leaf as the new sample, then the weight is the fraction of samples in the same leaf. If not, then the weight is zero. These weights for each y are summed up across all trees and averaged. Now since we have an array of target values and an array of weights corresponding to these target values, we can use this to measure empirical quantile estimates.

## References

There is a Python library extending Scitkit Learn for this method. The page here provides additional explanation.

https://scikit-garden.github.io/examples/QuantileRegressionForests/

**Comment by mhoward2718** [comment]
I'd be happy to try my hand at this with a few pointers to the relevant code sections

**Comment by guolinke** [comment]
Thanks @mhoward2718 
The RF code is in https://github.com/microsoft/LightGBM/blob/master/src/boosting/rf.hpp
And the current version of quantile Regression: 
https://github.com/microsoft/LightGBM/blob/6129208006c91f9ac94891690aca6f0c49d9b08e/src/objective/regression_objective.hpp#L464-L556

The core is in `RenewTreeOutput`, which takes the corresponding percentile values in each leaf.

**Comment by Bougeant** [comment]
Hi @guolinke,

I could not really figure out what RenewTreeOutput does, but I think @mhoward2718 was referring to a different technique to estimate the quantiles than the one which I believe is currently being used in LightBGM gradient boosting and random forest models, as well as in scikit-learn's gradient boosting model.

Indeed, and please correct me if I'm wrong, my understanding is that for both the GB and RF models, the value of the percentile of the samples associated to each leaf during training is stored in the model and then we take the average value of all the percentiles across all the leaves to make new inferences. 

For the scikit-garden RF model, all values of the all the training samples are recorded for each leave. To make new inferences, the distribution of all the training samples sharing a leaf with the inference samples is create and the selected percentile is selected from this distribution.

I've found that the scikit-garden approach leads to a better calibrated model (predicted quantiles correspond to true quantiles better) than the scikit-learn and lightGBM models for the datasets I've been using. It also leads to a lower quantile loss error for some datasets and more importantly, it guarantees that we can generate multiple consistent quantile predictions (i.e. correctly ranked), which is not the case for the lightGBM models.

Unfortunately the scikit-garden repo has been stalled for years now. Also the quantile=X in the .predict method is not consistent with the scikit-learn API which creates some difficulties for cross-validation.
